### ISSUE TITLE
[in_app_purchase] iOS: uncompleted purchase for non-consumable items doesn't come with queryPastPurchases.

### ISSUE THREAD

**Issue by benadrasni** [flutter/flutter#32759]
in_app_purchase: 0.1.0+4

I was testing in_app_purchase in sanbox on iOS. I did buy a non-consumable product, but I didn't call completePurchase on it. After that this purchase didn't come with queryPastPurchases and when I tried to buy it again, App Store didn't allow it because it was bought before. So now this purchase is stuck somewhere in the middle.

How can I reset it?

Doctor summary (to see all details, run flutter doctor -v):
[Ô^Ü] Flutter (Channel stable, v1.5.4-hotfix.2, on Microsoft Windows [Version
    6.1.7601], locale sk-SK)
[Ô^Ü] Android toolchain - develop for Android devices (Android SDK version 28.0.
3)
[Ô^Ü] Android Studio (version 3.4)
[!] IntelliJ IDEA Ultimate Edition (version 2019.1)
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.

**Comment by mklim** [comment]
You should be getting transaction updates to [the `purchaseUpdatedStream`](https://github.com/flutter/plugins/blob/master/packages/in_app_purchase/lib/src/in_app_purchase/in_app_purchase_connection.dart#L20-L39) in this case whenever the app opens. Are you seeing anything there?

/cc @cyanglaz

**Comment by cyanglaz** [comment]
Just to add on what @mklim has mentioned. You should get this purchase in the stream as soon as the app launches. as long as you subscribe to the stream early as mentioned in the guide.

**Comment by benadrasni** [comment]
I subscribed exactly like it is in the example. No transaction update is coming to the listener.

**Comment by mklim** [comment]
Got it, thanks for the info.

This is just a working theory, but I think you may be able to work around this for now by subscribing to the stream in `main()` itself instead of the top level widget's `initState` like in the example.

**Comment by rogiervandenberg** [comment]
I'm having the exact same issue. I tried relocating the stream subscription as described above. When I start the app, nothing happens on the `purchaseUpdatedStream`. Even when loading from `main()`.

In my shop screen: As soon as I hit the buy method, and update with `pending` is coming in on the stream. iOS prompts for buying. After clicking "Buy" on the prompt, iOS says the purchase has been made and will continue resuming my purchase. I click OK. No more new events coming in on the stream.

**Comment by cyanglaz** [comment]
@benadrasni It seems working for me using the example app. This is what I did.
1. I started with removing all the `completePurchase` code in the main.dart.
2. Then I run the app, buy an upgrade (non-consumable). Saw a pop-up saying purchase success. Record the `purchaseID`. And now i tried to buy the item again, i couldn't and Apple says already bought (so far it is all expected).
3. Close the app and run it again. Now I got an event from the `purchaseUpdatedStream` which contains the purchaseDetails with the same ID that I recorded from step 2, the status is `purchased` as expected. That means the previous purchase came back.
4. Now add back the removed `completePurchase` code and run the app again. Now the purchases were completed.

Note that the `completePurchase` should be called after you have a `purchaseDetails` object with `purchased` status. So when you buy an item without calling the `completePurchase`, the status of that purchase is still `purchased` by design. 

Please let us know if you can see this success scenario in our example app following my steps, or there is something that I am missing in my steps to reproduce.

**Comment by cyanglaz** [comment]
@rogiervandenberg If you are experiencing the same issue, could you also check my steps to reproduce works for you? I seems not been able to reproduce the issue.

**Comment by benadrasni** [comment]
@cyanglaz I didn't try example application, but I did all steps in my application without success. I am not getting event from the purchaseUpdatedStream even when I moved it to the main().

**Comment by cyanglaz** [comment]
@benadrasni Could you post a minimum reproducible code here?

**Comment by benadrasni** [comment]
@cyanglaz So I successfully run your example on my side. I disabled all completePurchase calls and successfully bought   "upgrade". I run app again and try buy it again (message "Already bought" appears). Then I put break point inside listener and nothing happend. I didn't get the event.

**Comment by cyanglaz** [comment]
You should get the event as soon as the app launches. Did you not get anything at all? Try add a print statement instead of break point to see if there are anything printed. You should be able to get some thing printed in the `onListen` function body.

**Comment by benadrasni** [comment]
@cyanglaz Break points deleted, print added to `listen` function body. No message was printed.

**Comment by benadrasni** [comment]
I am using iPhone 6 with iOS 12.2

**Comment by cyanglaz** [comment]
@benadrasni Have you found a solution to your issue?

**Comment by benadrasni** [comment]
@cyanglaz No, I didn't. I have created minimal application from your example and it didn't work. I kinda hope it won't appear at client side. If it will I'll be dealing with it case by case.

**Comment by cyanglaz** [comment]
@benadrasni ok. If you have any new discovery, please let us know. Or if you can provide a mini sample app to reproduce it that'd be great

**Comment by rav3r** [comment]
@cyanglaz I am facing the same problem, looks like InAppPurchasePlugin.handleTransactionsUpdated is called before SKPaymentQueueWrapper sets method call handler even if I subscribe to purchaseUpdatedStream in first main function line.

I think that you should somehow surface FIAPaymentQueueHandler.transactions so we can process transactions reported before purchaseUpdatedStream subscription.

**Comment by cyanglaz** [comment]
@rav3r Interesting. Thank you for providing the suggestion! Do you mind to wire up a PR, I will be happy to review it!

**Comment by rav3r** [comment]
@benadrasni I've found solution to my problem! My main function was async, I was awaiting loading stuff before calling runApp() so I delayed it's invocation a little bit - problem disappears when I've removed that delay. Looks like there is some setup code for streams inside runApp which is making streams work.

@cyanglaz I attached mini sample to reproduce this problem, you can simply remove "await Future.delayed(...)" to make it work again.

```
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:async';
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:in_app_purchase/in_app_purchase.dart';
import 'consumable_store.dart';

StreamSubscription<List<PurchaseDetails>> _subscription;

void main() async {
  print("SUBSCRIBE");
  Stream purchaseUpdated =
      InAppPurchaseConnection.instance.purchaseUpdatedStream;
  _subscription = purchaseUpdated.listen((purchaseDetailsList) {
    print("PURCHASE UPDATED");
  }, onDone: () {
    _subscription.cancel();
  }, onError: (error) {
    // handle error here.
  });

  await Future.delayed(const Duration(seconds: 1));

  runApp(MyApp());
}

const bool kAutoConsume = true;

const String _kConsumableId = 'consumable';
const List<String> _kProductIds = <String>[
  _kConsumableId
];

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final InAppPurchaseConnection _connection = InAppPurchaseConnection.instance;
  List<String> _notFoundIds = [];
  List<ProductDetails> _products = [];
  List<PurchaseDetails> _purchases = [];
  List<String> _consumables = [];
  bool _isAvailable = false;
  bool _purchasePending = false;
  bool _loading = true;
  String _queryProductError = null;

  @override
  void initState() {
    initStoreInfo();
    super.initState();
  }

  Future<void> initStoreInfo() async {
    final bool isAvailable = await _connection.isAvailable();
    if (!isAvailable) {
      setState(() {
        _isAvailable = isAvailable;
        _products = [];
        _purchases = [];
        _notFoundIds = [];
        _consumables = [];
        _purchasePending = false;
        _loading = false;
      });
      return;
    }

    ProductDetailsResponse productDetailResponse =
        await _connection.queryProductDetails(_kProductIds.toSet());
    if (productDetailResponse.error != null) {
      setState(() {
        _queryProductError = productDetailResponse.error.message;
        _isAvailable = isAvailable;
        _products = productDetailResponse.productDetails;
        _purchases = [];
        _notFoundIds = productDetailResponse.notFoundIDs;
        _consumables = [];
        _purchasePending = false;
        _loading = false;
      });
      return;
    }

    if (productDetailResponse.productDetails.isEmpty) {
      setState(() {
        _queryProductError = null;
        _isAvailable = isAvailable;
        _products = productDetailResponse.productDetails;
        _purchases = [];
        _notFoundIds = productDetailResponse.notFoundIDs;
        _consumables = [];
        _purchasePending = false;
        _loading = false;
      });
      return;
    }

    final QueryPurchaseDetailsResponse purchaseResponse =
        await _connection.queryPastPurchases();
    if (purchaseResponse.error != null) {
      // handle query past purchase error..
    }
    final List<PurchaseDetails> verifiedPurchases = [];
    for (PurchaseDetails purchase in purchaseResponse.pastPurchases) {
      if (await _verifyPurchase(purchase)) {
        verifiedPurchases.add(purchase);
      }
    }
    List<String> consumables = await ConsumableStore.load();
    setState(() {
      _isAvailable = isAvailable;
      _products = productDetailResponse.productDetails;
      _purchases = verifiedPurchases;
      _notFoundIds = productDetailResponse.notFoundIDs;
      _consumables = consumables;
      _purchasePending = false;
      _loading = false;
    });
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    List<Widget> stack = [];
    if (_queryProductError == null) {
      stack.add(
        ListView(
          children: [
            _buildConnectionCheckTile(),
            _buildProductList(),
            _buildConsumableBox(),
          ],
        ),
      );
    } else {
      stack.add(Center(
        child: Text(_queryProductError),
      ));
    }
    if (_purchasePending) {
      stack.add(
        Stack(
          children: [
            new Opacity(
              opacity: 0.3,
              child: const ModalBarrier(dismissible: false, color: Colors.grey),
            ),
            new Center(
              child: new CircularProgressIndicator(),
            ),
          ],
        ),
      );
    }

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('IAP Example'),
        ),
        body: Stack(
          children: stack,
        ),
      ),
    );
  }

  Card _buildConnectionCheckTile() {
    if (_loading) {
      return Card(child: ListTile(title: const Text('Trying to connect...')));
    }
    final Widget storeHeader = ListTile(
      leading: Icon(_isAvailable ? Icons.check : Icons.block,
          color: _isAvailable ? Colors.green : ThemeData.light().errorColor),
      title: Text(
          'The store is ' + (_isAvailable ? 'available' : 'unavailable') + '.'),
    );
    final List<Widget> children = <Widget>[storeHeader];

    if (!_isAvailable) {
      children.addAll([
        Divider(),
        ListTile(
          title: Text('Not connected',
              style: TextStyle(color: ThemeData.light().errorColor)),
          subtitle: const Text(
              'Unable to connect to the payments processor. Has this app been configured correctly? See the example README for instructions.'),
        ),
      ]);
    }
    return Card(child: Column(children: children));
  }

  Card _buildProductList() {
    if (_loading) {
      return Card(
          child: (ListTile(
              leading: CircularProgressIndicator(),
              title: Text('Fetching products...'))));
    }
    if (!_isAvailable) {
      return Card();
    }
    final ListTile productHeader = ListTile(
        title: Text('Products for Sale',
            style: Theme.of(context).textTheme.headline));
    List<ListTile> productList = <ListTile>[];
    if (!_notFoundIds.isEmpty) {
      productList.add(ListTile(
          title: Text('[${_notFoundIds.join(", ")}] not found',
              style: TextStyle(color: ThemeData.light().errorColor)),
          subtitle: Text(
              'This app needs special configuration to run. Please see example/README.md for instructions.')));
    }

    // This loading previous purchases code is just a demo. Please do not use this as it is.
    // In your app you should always verify the purchase data using the `verificationData` inside the [PurchaseDetails] object before trusting it.
    // We recommend that you use your own server to verity the purchase data.
    Map<String, PurchaseDetails> purchases =
        Map.fromEntries(_purchases.map((PurchaseDetails purchase) {
      if (Platform.isIOS) {
        InAppPurchaseConnection.instance.completePurchase(purchase);
      }
      return MapEntry<String, PurchaseDetails>(purchase.productID, purchase);
    }));
    productList.addAll(_products.map(
      (ProductDetails productDetails) {
        PurchaseDetails previousPurchase = purchases[productDetails.id];
        return ListTile(
            title: Text(
              productDetails.title,
            ),
            subtitle: Text(
              productDetails.description,
            ),
            trailing: previousPurchase != null
                ? Icon(Icons.check)
                : FlatButton(
                    child: Text(productDetails.price),
                    color: Colors.green[800],
                    textColor: Colors.white,
                    onPressed: () {
                      PurchaseParam purchaseParam = PurchaseParam(
                          productDetails: productDetails,
                          applicationUserName: null,
                          sandboxTesting: true);
                      if (productDetails.id == _kConsumableId) {
                        _connection.buyConsumable(
                            purchaseParam: purchaseParam,
                            autoConsume: kAutoConsume || Platform.isIOS);
                      } else {
                        _connection.buyNonConsumable(
                            purchaseParam: purchaseParam);
                      }
                    },
                  ));
      },
    ));

    return Card(
        child:
            Column(children: <Widget>[productHeader, Divider()] + productList));
  }

  Card _buildConsumableBox() {
    if (_loading) {
      return Card(
          child: (ListTile(
              leading: CircularProgressIndicator(),
              title: Text('Fetching consumables...'))));
    }
    if (!_isAvailable || _notFoundIds.contains(_kConsumableId)) {
      return Card();
    }
    final ListTile consumableHeader = ListTile(
        title: Text('Purchased consumables',
            style: Theme.of(context).textTheme.headline));
    final List<Widget> tokens = _consumables.map((String id) {
      return GridTile(
        child: IconButton(
          icon: Icon(
            Icons.stars,
            size: 42.0,
            color: Colors.orange,
          ),
          splashColor: Colors.yellowAccent,
          onPressed: () => consume(id),
        ),
      );
    }).toList();
    return Card(
        child: Column(children: <Widget>[
      consumableHeader,
      Divider(),
      GridView.count(
        crossAxisCount: 5,
        children: tokens,
        shrinkWrap: true,
        padding: EdgeInsets.all(16.0),
      )
    ]));
  }

  Future<void> consume(String id) async {
    await ConsumableStore.consume(id);
    final List<String> consumables = await ConsumableStore.load();
    setState(() {
      _consumables = consumables;
    });
  }

  void showPendingUI() {
    setState(() {
      _purchasePending = true;
    });
  }

  void deliverProduct(PurchaseDetails purchaseDetails) async {
    // IMPORTANT!! Always verify a purchase purchase details before delivering the product.
    if (purchaseDetails.productID == _kConsumableId) {
      await ConsumableStore.save(purchaseDetails.purchaseID);
      List<String> consumables = await ConsumableStore.load();
      setState(() {
        _purchasePending = false;
        _consumables = consumables;
      });
    } else {
      setState(() {
        _purchases.add(purchaseDetails);
        _purchasePending = false;
      });
    }
  }

  void handleError(IAPError error) {
    setState(() {
      _purchasePending = false;
    });
  }

  Future<bool> _verifyPurchase(PurchaseDetails purchaseDetails) {
    // IMPORTANT!! Always verify a purchase before delivering the product.
    // For the purpose of an example, we directly return true.
    return Future<bool>.value(true);
  }

  void _handleInvalidPurchase(PurchaseDetails purchaseDetails) {
    // handle invalid purchase here if  _verifyPurchase` failed.
  }

  static ListTile buildListCard(ListTile innerTile) =>
      ListTile(title: Card(child: innerTile));

  void _listenToPurchaseUpdated(List<PurchaseDetails> purchaseDetailsList) {
    purchaseDetailsList.forEach((PurchaseDetails purchaseDetails) async {
      if (purchaseDetails.status == PurchaseStatus.pending) {
        showPendingUI();
      } else {
        if (purchaseDetails.status == PurchaseStatus.error) {
          handleError(purchaseDetails.error);
        } else if (purchaseDetails.status == PurchaseStatus.purchased) {
          bool valid = await _verifyPurchase(purchaseDetails);
          if (valid) {
            deliverProduct(purchaseDetails);
          } else {
            _handleInvalidPurchase(purchaseDetails);
          }
        }
        if (Platform.isIOS) {
          InAppPurchaseConnection.instance.completePurchase(purchaseDetails);
        } else if (Platform.isAndroid) {
          if (!kAutoConsume && purchaseDetails.productID == _kConsumableId) {
            InAppPurchaseConnection.instance.consumePurchase(purchaseDetails);
          }
        }
      }
    });
  }
}

```

**Comment by cyanglaz** [comment]
@rav3r Thanks for the details! @benadrasni Do you want to take a look at the solution and see if that fixes for you?

**Comment by guopeng1994** [comment]
any progress ?

**Comment by benadrasni** [comment]
I tried a solution but it doesn't help. My main() is not async now but I didn't get the event.

**Comment by jonjon1123** [comment]
I am also experiencing the same issue as @benadrasni. I have tried the suggestions in this thread, but no luck. I can see the purchase come back in the stream as pending, but nothing after that. The purchase also doesn't show up when I use `queryPastPurchases` even though when I try to purchase the same item it says I have already purchased it so it will be restored (which also doesn't emit any events to the stream). FWIW, I am also doing all of this with a non-consumable product and it works just fine on Android.

**Comment by jonjon1123** [comment]
I have dove into the code for this plugin and can't seem to figure out why I don't get anything back from `queryPastPurchases`. It sure seems like there isn't anything being returned by the method channel here: https://github.com/flutter/plugins/blob/master/packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart#L125.

I pulled up my test iPhone's console output hoping to find some kind of indication of what is going on, but I didn't see anything. I am not too familiar with method channels. How can I troubleshoot this further?

**Comment by cyanglaz** [comment]
@jonjon1123 unfortunately, I couldn't reproduce the same issue. However, in order to help, I can list certain things that you can do to debug this.

To make sure the method channel method is called, you can open the xcode project (ios/xcworkspace). Look for `InAppPurchasePlugin.m`, put a break point here https://github.com/flutter/plugins/blob/8030246eaffeb89067455b6bf9b70ed1ce480773/packages/in_app_purchase/ios/Classes/InAppPurchasePlugin.m#L210
And run the app using xcode(You have to use xcode to hit the break point, do not use flutter run), execute the query past purchase. You should be able to hit the break point this way and get some details. Note that this method won't return anything directly, one of the observing method (such as https://github.com/flutter/plugins/blob/8030246eaffeb89067455b6bf9b70ed1ce480773/packages/in_app_purchase/ios/Classes/FIAPaymentQueueHandler.m#L67)should receive a notification, you might want to put break points to the start of each observing method. 

Let me know if you face any troubles during debugging please. I appreciate if you could help debugging this!

**Comment by jonjon1123** [comment]
Ok, I finally got my mac back up and running. I followed the steps provided by @cyanglaz (thanks!!) and it seems the `restoreTransactions` iOS method is being called when I call `queryPastPurchases` in the Flutter code, but the breakpoint in the `paymentQueue` method on iOS isn't being hit (even though `restoreTransactions` is). The only time I can get the breakpoint in the `paymentQueue` method to hit is when I first launch the app. When it is hit It looks like there are items in the `transactions` array, but I can't see what they are in the XCode debugger (see screenshot below). I am not too familiar with XCode so maybe I am missing something...

![image](https://user-images.githubusercontent.com/691969/62416849-f9979d00-b607-11e9-9e09-c766b39faf02.png)

So I guess my questions are:
1. How can I see the values of what is in the `transactions` array?
2. Why is the  `paymentQueue` iOS method not called when `restoreTransactions` iOS method is called?

**Comment by cyanglaz** [comment]
@jonjon1123 
> 1. How can I see the values of what is in the transactions array?

In your screenshot, on the left bottom of the window, if you click the arrow to expand the SKPaymentTransaction object, you should be able to see all the fields in there

> 2. Why is the paymentQueue iOS method not called when restoreTransactions iOS method is called?

This is weird.
When the restoreTransaction gets called, did it go to the applicationName route or non-applicationName route. 
Also, maybe you can try to put breakpoints to all the observing methods to see if other methods get triggered such as `restoreCompletedTransactionsFailedWithError`, then there might be an error and you can see what error it is.
All the observing methods should be under `#pragma mark - observing`

**Comment by jonjon1123** [comment]
> if you click the arrow to expand the SKPaymentTransaction object, you should be able to see all the fields in there

The screenshot has all of the items in the tree under `SKPaymentTransaction` expanded, but I don't see any fields for the `SKPaymentTransaction` object actually being shown. It is like it is only showing references or something.

> When the restoreTransaction gets called, did it go to the applicationName route or non-applicationName route.

It goes the applicationName route which is expected as I am providing an applicationName.

> Also, maybe you can try to put breakpoints to all the observing methods

I did this and the only other method that gets called is `paymentQueueRestoreCompletedTransactionFinished`, but no items are returned on the `purchaseUpdatedStream`.

One other piece of information... I was thinking the issue might have something to do with my code not subscribing to the `purchaseUpdatedStream` soon enough. So I put the following as the very first block of code in my apps main method:
```
StreamSubscription subscription = InAppPurchaseConnection.instance.purchaseUpdatedStream.listen((List<PurchaseDetails> purchases) {
    purchases.forEach((PurchaseDetails purchaseDetails) {
      print('PURCHASE (listen): '+purchaseDetails.productID+' | '+purchaseDetails.status.toString());
    });
  })..onError((dynamic e) => print(e))
  ..onDone(() => print('DONE'));
```
When I run the app through `flutter run` I don't see anything logged out, BUT when I run it through XCode I see the `PURCHASE (listen): ...` log items when the app first starts and they seem to be printing out the previous purchases I am looking for! With that said, those log items don't show either way (`flutter run` or XCode) if I try to `queryPastPurchases`.

**Comment by jonjon1123** [comment]
Another update on this...

I am really in need of getting this resolved to move forward so for the time being I switched to using https://pub.dev/packages/flutter_inapp_purchase. I know it is being deprecated, but I need somthing that works. I was also curious if I would see the same results. I created a branch at the state I mentioned above (so I could continue troubleshooting) and did the changes to convert to the new package in a different branch. After wiring everything up with the new package, I saw the exact same issue with the existing purchases (the ones I had done in the past) not being returned on iOS. I then tried setting up a brand new tester account and went through the purchase flow from the beginning (since the other ones had already done that using the previous lib). After the flow was done (through the new lib), I received the purchase I just did with the new tester account when I called for previous purchases (but not the others)!

So it seems there must be something about how these past purchases where made in iOS that is preventing them from returning (since both packages didn't return them).

Some thoughts:

- Maybe the issue has something to do with the applicationName? The other package doesn't support this so it wasn't set.
- I had issues with finishing the transactions in iOS when I was first deving this and I know the ones that aren't returning never got finished properly. Maybe that is the issue? With that said, if I try to purchase the product again I get an error saying it is already purchased and will be restored.

**Comment by cyanglaz** [comment]
@jonjon1123 it sounds like the issue has something to do with the test account that you previously used.

May I ask what you used for the applicationName. The applicationName should be some encoded user name for your application.
For example userA signed up for your application, they used some user name, e.g, purpleFlower123. 
The application user name for this user using your purchase should be some encrypted string with represents `purpleFlower123` (Although I think the name `applicationName` is a bit confusing, we might need to update it to `applicationUserName`)

Also, do you mean that everything works fine if you use a new testing account?

on the other note, I know that iOS doesn't support remove purchase from a testing account. So if you bought a non-consumable using a testing account, you won;'t be able to purchase it again ever. you will need to create another test account to test.

**Comment by jonjon1123** [comment]
> May I ask what you used for the applicationUserName.
I use their email address that has been Salsa20 encrypted and then converted to base64.

> Also, do you mean that everything works fine if you use a new testing account?
I had tested a brand new tester account in the past and it didn't work. This time, I tried removing sending the applicationUserName and it worked!

So I guess the question now is, why was the way I was setting the applicationUserName causing this issue?

**Comment by cyanglaz** [comment]
@jonjon1123 The only thing I can think is that the encrypted string that you used when purchasing is different from the encrypted string you used when querying the past purchases. Maybe confirm that?

**Comment by jonjon1123** [comment]
@cyanglaz I logged out the encrypted string that is used for `queryPastPurchases` and `buyNonConsumable` and it is exactly the same.

**Comment by ziggycrane** [comment]
I am having the same issue. I can purchase the subscription get the success dialog, also InAppPurchaseConnection.instance.purchaseUpdatedStream gets called, but when I try to InAppPurchaseConnection.instance.queryPastPurchases() it returns me an empty array in pastPurchases array. 

When I try to purchase again, it warns me that I have already purchased. 

Restarting app/phone and anything else I have tried does not work, queryPastPurchases always returns an empty array. If I remember correctly, it didn't when I bought a simple non-consumable product on another project I worked on, but it does when I purchase a subscription. 
Don't know if that helps to find a solution.

**Comment by qq326646683** [comment]
maybe this can help you
https://github.com/qq326646683/plugins/commit/7cbe66572ae72c0a5f7baacb21194581b61acb74

**Comment by JarvanMo** [comment]
Facing the same problem because I forget to complete purchase. Then the apple never allow me to purchase again.

**Comment by JarvanMo** [comment]
As @qq326646683 mentioned, it's a quick fix for this problem.

**Comment by singlebyted** [comment]
Same issue, @qq326646683 's solution is worked for me.

**Comment by MasiaVittorio** [comment]
Sorry, can anyone explain to me how to implement and use @qq326646683's solution?
How do I install his version of the plugin with the fix? How should I then use that added method?

Thank you and sorry again for the dumb question

**Comment by qq326646683** [comment]
@MasiaVittorio you can add it into pubspec.yaml
```
in_app_purchase:
    git:
      url: https://github.com/qq326646683/plugins.git
      ref: 13df320b6112a3a4abfbec47bba53b2f95402637
      path: packages/in_app_purchase
```

then you can refer to this Chinese [document](https://juejin.im/post/5dd36a59518825438838212a)

**Comment by ThinkSimple** [comment]
I have a similar problem, whenever I call queryPastPurchases() on ios I get an empty list back. But 
the app store says that I have already bought this item.

**Comment by miikaeru** [comment]
I have the same issue. I purchased a non-consumable product on iOS in the sandbox with a testuser, but  queryPastPurchases() returns an empty list.

**Comment by Vardiak** [comment]
Any news on this issue ? Non-completed transactions should be queued, before we subscribe to the stream...

**Comment by andr81** [comment]
@qq326646683 Thanks! It works. Is it possible to contribute to the main repository?

**Comment by agordeev** [comment]
I'm getting the same. It's actually a serious issue: we're unable to complete a purchase that hang!
@qq326646683 solution works, but his version of the package is outdated. Can his code be merged into the repo? @mklim

**Comment by lalawila** [comment]
same problem

**Comment by jacksonw765** [comment]
Can confirm that @qq326646683 fix still works for anyone facing queryPastPurchases() returning an empty string. I think the issue is that you MUST mark them as completed in iOS otherwise your PurchaseDetails object will be trapped in a completed state but not able to be retrieved like what happened with many of us. I've been fighting this issue for the last 4 days lol.

**Comment by LHLL** [comment]
It should be fixed in version: 0.3.2+1

**Comment by jacksonw765** [comment]
I'm actually having more issues using branch 0.3.2+1. Android is unable to fetch any products (was working on qq326646683 fix from December) and iOS keeps saying "There is a pending transaction for the same product identifier...".  iOS was working fine other than products getting caught in that non completed state.

**Comment by LHLL** [comment]
I see, you already have multiple incomplete transactions in the queue, the fix was intended to prevent it from happening. 

To unblock you from current state you can follow these steps
1. Find following method defined in [InAppPurchasePlugin.m](https://github.com/flutter/plugins/blob/master/packages/in_app_purchase/ios/Classes/InAppPurchasePlugin.m#L279) and replace the implementation:
```
- (void)handleTransactionsUpdated:(NSArray<SKPaymentTransaction *> *)transactions {
    // This will remove all old transactions regardless of states.
    for (SKPaymentTransaction *transaction in transactions) {
        [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
    }
}
```
2. Launch your app on the test device one.
3. Restore the implementation of the above method:
```
- (void)handleTransactionsUpdated:(NSArray<SKPaymentTransaction *> *)transactions {
  NSMutableArray *maps = [NSMutableArray new];
  for (SKPaymentTransaction *transaction in transactions) {
    [maps addObject:[FIAObjectTranslator getMapFromSKPaymentTransaction:transaction]];
  }
  [self.callbackChannel invokeMethod:@"updatedTransactions" arguments:maps];
}
```
4. Add a new transaction and verify.

**Comment by benadrasni** [comment]
@LHLL  please reopen this issue. It is even worse that it has been. I tested version 0.3.2+1 and if for whatever reason user cancel purchase in iOS he/she won't be able to purchase same product again. When I switched back to version 0.3.1+2 I was able to purchase product which was previously canceled. I think the problem is with transactions which are coming with error. They need to be removed from the queue.

**Comment by jacksonw765** [comment]
@benadrasni I agree I was having that issue as well even with the fix LHLL provided. I ended up getting very unpredictable results. I ended up switching to purchases_flutter by revenue cat which worked much better.

**Comment by LHLL** [comment]
@benadrasni 
It's an error state if a failed transaction with the same product identifier is in the queue while the user is trying to purchase the same product. It can cause the subscriptions being delivered to the wrong user if the app supports third-party login such as login with Google.

As a library, this plugin cannot remove/finish any transactions for the client since some clients need to complete extra logic before finishing a transaction, thus, it's the client app's responsibility to remove failed transactions as needed. 

The safe flow should be:
1. User wants to purchase product with product identifier A,
2. Client app checks whether or not the payment queue contains a transaction with the same product identifier.
3. If yes, then let the user know a pending transaction is there.
4. If no, then add the transaction to the queue.

```
var paymentWrapper = SKPaymentQueueWrapper();
var transactions = await paymentWrapper.transactions();
transactions.forEach((transaction) async {
   await paymentWrapper.finishTransaction(transaction);
});
```

This is due to the technical limitations from Apple's side:
1. Apple doesn't provide a transaction identifier until a transaction is purchased or failed.
2. Apple doesn't expose a user's current Apple ID to a third-party app.

Having two transactions of the same products can be really tricky for delivering the product to the right user since user can switch to a different account at any time. And if your app happen to provide auto-renewable subscription, then a transaction can be added to the payment queue at any time without user interaction,

**Comment by benadrasni** [comment]
OK. So what do I have to do when user initiate a purchase, but cancel it in
iOS? I will get transaction with error to the listener. What should I call
to remove this transaction from the queue? Please advise and update an
example for such a case. Thanks.

**Comment by LHLL** [comment]
@benadrasni Sure, does following example answer your questions?
```
class PurchaseService{
    final _connection = InAppPurchaseConnection();

    const PurchaseService(){
          _connect.listen((detail) {
               ...
              // Say you already know it's a cancelled transaction here
              _connect.completePurchase(detail);
          });    
    }

    void purchase({your purchase params}) async {
         await _connection. buyConsumable({your custom params});
    }
}
```

**Comment by poqueque** [comment]
There is another case where the user closes the app or simply does not finish the purchase after initiating it (user can run out of battery, lost internet connection...). In that case there is no way to complete the unfinished purchase after app restart, so the user will never be able to do the purchase again.

How can this case be managed?

**Comment by LHLL** [comment]
@poqueque As the documentation suggests, subscribe to the stream as soon as app launches:
https://github.com/flutter/plugins/blob/master/packages/in_app_purchase/lib/src/in_app_purchase/in_app_purchase_connection.dart#L4

```
class PurchaseService{
    final _connection = InAppPurchaseConnection();

    const PurchaseService(){
          _connection.purchaseUpdatedStream((detail) {
               ...
              // Verify the receipt for succeeded transactions and handle the error for the failed
              // transaction here.
              _connection.completePurchase(detail);
          });    
    }
}
```

**Comment by benadrasni** [comment]
So I tested latest (0.3.4) version. I was able to complete purchase when it
was cancelled in iOS, but when I initiate purchase and iOS screen with Buy
and Cancel option appears I switch off the mobile phone. Then I switch on
it again and run the application in debug mode.

I was expecting that initiated purchase will appear in result of
InAppPurchaseConnection.instance.queryPastPurchases() but it does not. So I
was not able to complete this "bad" purchase and next time I tried to
initiate it, duplicate purchase error appears so I am not able to do this
purchase again.

Please return uncompleted purchases with error with queryPastPurchases()

Thanks

**Comment by kaushaldeligence** [comment]
> @benadrasni
> It's an error state if a failed transaction with the same product identifier is in the queue while the user is trying to purchase the same product. It can cause the subscriptions being delivered to the wrong user if the app supports third-party login such as login with Google.
> 
> As a library, this plugin cannot remove/finish any transactions for the client since some clients need to complete extra logic before finishing a transaction, thus, it's the client app's responsibility to remove failed transactions as needed.
> 
> The safe flow should be:
> 
> 1. User wants to purchase product with product identifier A,
> 2. Client app checks whether or not the payment queue contains a transaction with the same product identifier.
> 3. If yes, then let the user know a pending transaction is there.
> 4. If no, then add the transaction to the queue.
> 
> ```
> var paymentWrapper = SKPaymentQueueWrapper();
> var transactions = await paymentWrapper.transactions();
> transactions.forEach((transaction) async {
>    await paymentWrapper.finishTransaction(transaction);
> });
> ```
> 
> This is due to the technical limitations from Apple's side:
> 
> 1. Apple doesn't provide a transaction identifier until a transaction is purchased or failed.
> 2. Apple doesn't expose a user's current Apple ID to a third-party app.
> 
> Having two transactions of the same products can be really tricky for delivering the product to the right user since user can switch to a different account at any time. And if your app happen to provide auto-renewable subscription, then a transaction can be added to the payment queue at any time without user interaction,

this worked for me

**Comment by LHLL** [comment]
> So I tested latest (0.3.4) version. I was able to complete purchase when it was cancelled in iOS, but when I initiate purchase and iOS screen with Buy and Cancel option appears I switch off the mobile phone. Then I switch on it again and run the application in debug mode. I was expecting that initiated purchase will appear in result of InAppPurchaseConnection.instance.queryPastPurchases() but it does not. So I was not able to complete this "bad" purchase and next time I tried to initiate it, duplicate purchase error appears so I am not able to do this purchase again. Please return uncompleted purchases with error with queryPastPurchases() Thanks
> […](#)

Some more context here:
`queryPastPurchases()` on iOS platform maps to [restoring purchased products](https://developer.apple.com/documentation/storekit/in-app_purchase/restoring_purchased_products?language=objc). Past purchases on iOS means a transaction has been finished, if `finishTransaction` is not called, then this transaction is still an open transaction that needs extra steps to process(logging, verify receipt, etc), thus cannot be picked up by restoring transactions method.

**Comment by erf** [comment]
As i commented [here](https://github.com/flutter/flutter/issues/53534#issuecomment-640223358), the plugin `flutter_inapp_purchase` has a method `clearTransactionsIOS` which solved my problem. I included this package and called this method before purchasing a new non-consumable / subscription. Might this solve your problem?

**Comment by Estejin** [comment]
The problem has not be solved int the latest version. queryPastPurchases() return a empty list.
How to implement the Restore procedure for iOS?
If you use No-Cosumable item, this plugin is recommended at all now.

**Comment by kofj** [comment]
> Restore
Same issue too, queryPastPurchases() return nothing at last version `in_app_purchase: ^0.3.4+15`.

**Comment by lehoai** [comment]
2021, I'm having the exact same issue. lol
